src/crush/CrushCompiler.cc

```diff
@@ -816,6 +816,48 @@ void CrushCompiler::dump(iter_t const& i, int ind)
 }
 
 
+//Sometimes when defined the bucket above the item which the bucket
+//included.It need to move the item above the bucket.
+//Otherwise the following analyse will error for can not find the item.
+int CrushCompiler::adjust_bucket_item_place(iter_t const &i)
+{
+  map<string,set<string> > bucket_items;
+  map<string,iter_t> bucket_itrer;
+  vector<string> buckets;
+  for (iter_t p = i->children.begin(); p != i->children.end(); ++p) {
+    if ((int)p->value.id().to_long() == crush_grammar::_bucket) {
+      string name = string_node(p->children[1]);
+      buckets.push_back(name);
+      bucket_itrer[name] = p;
+      //skip non-bucket-item children in the bucket's parse tree
+      for (unsigned q=3; q < p->children.size()-1; ++q) {
+        iter_t sub = p->children.begin() + q;
+        if ((int)sub->value.id().to_long() == crush_grammar::_bucket_item) {
+          string iname = string_node(sub->children[1]);
```

Comment:
Here is very strange.Use the command git show --color it does not look the same as here.

+        if ((int)sub->value.id().to_long() == crush_grammar::_bucket_item) {
+          string iname = string_node(sub->children[1]);
+          bucket_items[name].insert(iname);
+        }